pkg/gce-pd-csi-driver/controller.go

```diff
@@ -565,3 +574,127 @@ func getZoneFromSegment(seg map[string]string) (string, error) {
+	}
+	remainingZones := sets.NewString(totZones...).Difference(sets.NewString(existingZones...))
+	l := remainingZones.List()
+	if len(l) < needToGet {
+		return nil, fmt.Errorf("not enough remaining zones in %v to get %v zones out", l, needToGet)
+	}
+	// add l and zones
+	ret := append(existingZones, l[0:needToGet]...)
+	if len(ret) != numZones {
+		return nil, fmt.Errorf("did some math wrong, need %v zones, but got %v", numZones, ret)
+	}
+	return ret, nil
+}
+
+func generateCreateVolumeResponse(selfLink string, capBytes int64, zones []string) *csi.CreateVolumeResponse {
+	tops := []*csi.Topology{}
+	for _, zone := range zones {
+		tops = append(tops, &csi.Topology{
+			Segments: map[string]string{common.TopologyKeyZone: zone},
+		})
+	}
+	createResp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			CapacityBytes:      capBytes,
+			Id:                 cleanSelfLink(selfLink),
+			Attributes:         nil,
+			AccessibleTopology: tops,
+		},
+	}
+	return createResp
+}
+
+func cleanSelfLink(selfLink string) string {
+	temp := strings.TrimPrefix(selfLink, gce.GCEComputeAPIEndpoint)
+	return strings.TrimPrefix(temp, gce.GCEComputeBetaAPIEndpoint)
+}
+
+func createRegionalDisk(ctx context.Context, cloudProvider gce.GCECompute, name string, zones []string, diskType string, capacityRange *csi.CapacityRange, capBytes int64) (*gce.CloudDisk, error) {
+	region, err := common.GetRegionFromZones(zones)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get region from zones: %v", err)
+	}
+
+	fullyQualifiedReplicaZones := []string{}
+	for _, replicaZone := range zones {
+		fullyQualifiedReplicaZones = append(
+			fullyQualifiedReplicaZones, cloudProvider.GetReplicaZoneURI(replicaZone))
+	}
+
+	err = cloudProvider.InsertDisk(ctx, meta.RegionalKey(name, region), diskType, capBytes, capacityRange, fullyQualifiedReplicaZones)
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert regional disk: %v", err)
+	}
+
+	glog.V(4).Infof("Completed creation of disk %v", name)
+	disk, err := cloudProvider.GetDisk(ctx, meta.RegionalKey(name, region))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get disk after creating regional disk: %v", err)
+	}
+	glog.Warningf("GCE PD %s already exists after wait, reusing", name)
+	return disk, nil
+}
+
+func createSingleZoneDisk(ctx context.Context, cloudProvider gce.GCECompute, name string, zones []string, diskType string, capacityRange *csi.CapacityRange, capBytes int64) (*gce.CloudDisk, error) {
+	if len(zones) != 1 {
+		return nil, fmt.Errorf("got wrong number of zones for zonal create volume: %v", len(zones))
+	}
+	diskZone := zones[0]
+	err := cloudProvider.InsertDisk(ctx, meta.ZonalKey(name, diskZone), diskType, capBytes, capacityRange, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert zonal disk: %v", err)
+	}
+
+	glog.V(4).Infof("Completed creation of disk %v", name)
+	disk, err := cloudProvider.GetDisk(ctx, meta.ZonalKey(name, diskZone))
```

Comment:
now for each creation, there is an extra cloud provider GET API calls. Why not just use insertDisk which can return the disk object directly? In previous version, there is no GET call after the CREATE call, right?

Reply:
insertDisk doesn't return a disk object AFAIK. It returns an operation which we can poll for completion of the disk create operation, but getting the disk itself will be a seperate operation still.

We required this change not strictly because of RePD but because we wanted to have `volume ID` be the disk `selfLink` of the form `projects/{project}/zones|regions/{zone|region}/disks/{diskname}` and to get the self-link we need to get the disk